src/fairseq2/recipes/jepa/models.py

```diff
@@ -0,0 +1,320 @@
+
+        # Note that the cross-attention norm is applief on encoder output and not seqs
+        if self.norm_order != TransformerNormOrder.POST:
+            encoder_output = self.cross_attn_layer_norm(encoder_output)
+
+        seqs = self.cross_attn(
+            seqs,
+            padding_mask,
+            keys=encoder_output,
+            key_padding_mask=encoder_padding_mask,
+            values=encoder_output,
+            state_bag=state_bag,
+        )
+
+        if self.cross_attn_dropout is not None:
+            seqs = self.cross_attn_dropout(seqs)
+
+        seqs = self.cross_attn_residual(seqs, residual)
+
+        if self.norm_order == TransformerNormOrder.POST:
+            seqs = self.cross_attn_layer_norm(seqs)
+
+        return seqs
+
+    def _forward_ffn(self, seqs: Tensor) -> Tensor:
+        residual = seqs
+
+        if self.norm_order != TransformerNormOrder.POST:
+            seqs = self.ffn_layer_norm(seqs)
+
+        seqs = self.ffn(seqs)
+
+        if self.ffn_dropout is not None:
+            seqs = self.ffn_dropout(seqs)
+
+        seqs = self.ffn_residual(seqs, residual)
+
+        if self.norm_order == TransformerNormOrder.POST:
+            seqs = self.ffn_layer_norm(seqs)
+
+        return seqs
+
+    def extra_repr(self) -> str:
+        """:meta private:"""
+        s = super().extra_repr()
+
+        return f"{s}, model_dim={self.model_dim}, norm_order={self.norm_order.name}"
+
+
+@final
+class AttentivePooler(Module):
+    """
+    An attentive pooler that gets output of a Jepa encoder and decode it into
+    a logit of a given task.
+
+    TODO:
+    - Move this into fairseq2.nn to benefit other similiar tasks. Internally,
+    this module is just a thin transformer encoder without self attention layer.
+    Optionally, it can consist of some extra transformer encoders depending on the
+    (finetuning) task
+    """
+
+    model_dim: int
+    num_pools: int
+    decoder: CrossAttentionDecoder
+    encoder: TransformerEncoder | None
+    init_fn: Callable[[Tensor], None] | None
+
+    def __init__(
+        self,
+        decoder: CrossAttentionDecoder,
+        encoder: TransformerEncoder | None,
+        *,
+        num_pools: int = 1,
+        init_fn: Callable[[Tensor], None] | None = None,
+        device: Device | None = None,
+        dtype: DataType | None = None,
+    ) -> None:
+        super().__init__()
+
+        self.model_dim = decoder.model_dim
+
+        self.decoder = decoder
+
+        if encoder:
+            self.encoder = encoder
+        else:
+            self.register_module("encoder", None)
+
+        self.num_pools = num_pools
+        self.pool_layer = Parameter(
+            torch.empty(1, num_pools, self.model_dim, device=device, dtype=dtype)
+        )
+
+        if init_fn:
+            init_fn(self.pool_layer)
+
+    def forward(
+        self, seqs: Tensor, padding_mask: PaddingMask | None
+    ) -> tuple[Tensor, PaddingMask | None]:
+        if self.encoder:
```

Comment:
Are we sure that this forward implementation is accurate? In the reference implementation [here](https://github.com/facebookresearch/jepa/blob/main/src/models/attentive_pooler.py#L97-L102), I see that encoder stack is applied *after* cross attention which is the reverse of what we have here.

Reply:
the internal code seems to deviate from this